storage/imx6ull-flash/imx6ull-flash-test.c

```diff
@@ -112,6 +114,164 @@ void print_block(const uint8_t *data, unsigned int len)
+		return -ENOMEM;
+
+	if ((data = malloc(BLOCK_PAGES_CNT * sizeof(uint8_t *))) == NULL) {
+		flashdrv_dmadestroy(dma);
+		return -ENOMEM;
+	}
+
+	do {
+		/* Read raw block data into memory and flip bits */
+		for (i = 0; i < BLOCK_PAGES_CNT; i++) {
+			if ((data[i] = mmap(NULL, (FLASHDRV_PAGESZ + _PAGE_SIZE - 1) & ~(_PAGE_SIZE - 1), PROT_READ | PROT_WRITE, MAP_UNCACHED, OID_CONTIGUOUS, 0)) == MAP_FAILED) {
+				err = -ENOMEM;
+				printf("mmap() failed: %d\n", err);
+				break;
+			}
+
+			if ((err = flashdrv_readraw(dma, paddr + i, data[i], FLASHDRV_PAGESZ)) < 0) {
+				printf("readraw() failed: %d\n", err);
+				break;
+			}
+
+			for (offs = i * FLASHDRV_PAGESZ; start < min(end, offs + FLASHDRV_PAGESZ); start += (end - start) / n--)
+				data[i][start - offs] ^= 1 << rand() % 8;
+		}
+
+		if (err)
+			break;
+
+		/* Erase the block */
+		if ((err = flashdrv_erase(dma, paddr)) < 0) {
+			printf("erase() failed: %d\n", err);
+			break;
+		}
+
+		/* Write back raw block data with flipped bits */
+		for (j = 0; j < BLOCK_PAGES_CNT; j++) {
+			if ((err = flashdrv_writeraw(dma, paddr + j, data[j], FLASHDRV_PAGESZ)) < 0) {
+				printf("writeraw() failed: %d\n", err);
+				break;
+			}
+
+			/* Check ECC errors */
+			aux = (flashdrv_meta_t *)(data[j] + flashinfo.writesz);
+			if ((err = flashdrv_read(dma, paddr + j, data[j], aux)) < 0) {
+				printf("read() failed: %d\n", err);
+				break;
+			}
+
+			for (k = 0; k < sizeof(aux->errors); k++) {
+				if (aux->errors[k] && (aux->errors[k] != 0xff)) {
+					printf("[%4u]: ", paddr + j);
+
+					if (aux->errors[k] == 0xfe)
+						printf("uncorrectable data in chunk %u\n", k);
+					else
+						printf("%u corrections in chunk %u\n", aux->errors[k], k);
+				}
+			}
+		}
+	} while (0);
+
+	for (j = 0; j < i; j++)
+		munmap(data[j], (FLASHDRV_PAGESZ + _PAGE_SIZE - 1) & ~(_PAGE_SIZE - 1));
+	free(data);
+	flashdrv_dmadestroy(dma);
+
+	return err;
+}
+
+
+/* Breaks ECC correction / sets BBM in FCB, DBBT, FW1 and FW2 */
+/* (after running the test the BootROM shouldn't be able to boot from NAND flash) */
+void test_bootrom(void)
+{
+	flashdrv_dma_t *dma;
+
+	dma = flashdrv_dmanew();
+
+	do {
+		/* FCB structure encoded with BCH40 (as expected by the BootROM) */
+		/* [32B]  8x[128B + 65B]   [536B] */
+		/* [META] 8x[DATA + ECC40] [SPARE] */
+
+		/* Break FCB0 first data chunk (ECC should correct up to 40 bit flips) */
+		if (flip_bits(41, 0, 32, 32 + 128 + 65) < 0) {
+			printf("failed to break FCB0\n");
+			break;
+		}
+
+#if 0
+		/* Set FCB0 BBM */
+		if (flashdrv_markbad(dma, 0 * 64) < 0) {
+			printf("failed to set FCB0 BBM\n");
+			break;
+		}
+#endif
+
+		/* Default page structure (only FCB is read with different configuration by the BootROM) */
+		/* [16B + 26B]    8x[512B + 22.75B] [32B] */
+		/* [META + ECC16] 8x[DATA + ECC14]  [SPARE] */
+
+		/* Break DBBT0 first data chunk (ECC should correct up to 14 bit flips) */
+		if (flip_bits(15, 4, 16 + 26, 16 + 26 + 512 + 22) < 0) {
```

Comment:
maybe we can specify some of these magic values as eg. `const unsigned int rawmetasz = 16 + 26`, so the test definitions would be more semantic?

Reply:
Done.